CI: Build Embedded Boundaries

ax3l opened the pull request:
## Summary

- [x] Add coverage for EB in CI
- [x] fix DPC++ single precision build
- [x] ~~fix `LinearSolvers` array bounds w/ GCC 7.5~~ -> false positive

## Additional background

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

ax3l committed:
CI: Build Embedded Boundaries

Add coverage for EB.

WeiqunZhang commented:
Do we want some tests that do not use EB?  It's possible to have errors that appear only when EB is not enabled.

ax3l committed:
Fix: copysign() for Single Precision

Seen with DPC++ in WarpX:
```
/home/runner/work/WarpX/WarpX/build_sp/_deps/fetchedamrex-src/Src/EB/AMReX_EBMultiFabUtil_3D_C.H:256:43: error: no matching function for call to 'copysign'
            int jj = j + static_cast<int>(amrex::Math::copysign(1.0, fcx(i,j,k,0)));
                                          ^~~~~~~~~~~~~~~~~~~~~
/opt/intel/oneapi/compiler/2021.3.0/linux/bin/../include/sycl/CL/sycl/builtins.hpp:110:55: note: candidate template ignored: deduced conflicting types for parameter 'T' ('double' vs. 'float')
detail::enable_if_t<detail::is_genfloat<T>::value, T> copysign(T x,
                                                      ^
...
```

ax3l commented:
> Do we want some tests that do not use EB? It's possible to have errors that appear only when EB is not enabled.

Good idea - in WarpX I add a single CI test that builds a "minimal" build - disabling as much features as possible.
That way, we can cover that we can truly disable features at compile-time.

ax3l commented:
@WeiqunZhang can you take a look at the `-Werror=array-bounds` error in `Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp`?
```
Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp
In file included from /home/runner/work/amrex/amrex/Src/Base/AMReX_ArrayLim.H:9:0,
                 from /home/runner/work/amrex/amrex/Src/Base/AMReX_Box.H:7,
                 from /home/runner/work/amrex/amrex/Src/Base/AMReX_FabFactory.H:6,
                 from /home/runner/work/amrex/amrex/Src/EB/AMReX_EBFabFactory.H:5,
                 from /home/runner/work/amrex/amrex/Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.H:5,
                 from /home/runner/work/amrex/amrex/Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp:2:
/home/runner/work/amrex/amrex/Src/Base/AMReX_Array.H: In member function ‘virtual void amrex::MLEBABecLap::applyBC(int, int, amrex::MultiFab&, amrex::MLLinOp::BCMode, amrex::MLLinOp::StateMode, const amrex::MLMGBndry*, bool) const’:
/home/runner/work/amrex/amrex/Src/Base/AMReX_Array.H:114:23: error: array subscript is above array bounds [-Werror=array-bounds]

WeiqunZhang commented:
I suspect this is false positive.  GCC's `-Werror=array-bounds` has a lot of false positives.  I could not reproduce it on my desktop with gcc 9.3.0.  May have to use gcc 7.5.

ax3l committed:
EB Test: Fix extra semicolon

ax3l committed:
CI: EB not in CUDA 9.2

```
CMake Error at Src/CMakeLists.txt:84 (message):
  EB component of AMReX is not compatible with CUDA 9.2
```

ax3l commented:
Thanks for the check, see it now as well in the other entries.
I would say for the sake of compiler coverage and since we have also newer GCCs in CI, I would add `-Wno-array-bounds` to the one GCC 7.5 build that has the false positive.

Alternatively, I can also just disable EB in that one build again, to keep the very first build small and minimal. I pushed that for now.

What do you think?

ax3l committed:
CI: Keep FIrst Entry Simple

Avoids the GCC 7.5 false-positive in `-Warray-bounds`, too.

WeiqunZhang commented:
Sounds good.  When I get time, I will install gcc 7.5 and figure out what it complains about.

ax3l commented:
Looking where this is coming from:
```
36/38 Test #36: LinearSolvers_NodalPoisson ..........***Failed    0.72 sec
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (21.08.0) initialized
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
SIGABRT
See Backtrace.0 file for details
Abort(6) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 6) - process 0
```

WeiqunZhang commented:
Possible reason is that the non-EB test does work when EB is enabled.

WeiqunZhang commented:
I can push a fix to this PR.

WeiqunZhang committed:
avoid bad_cast

ax3l commented:
I think the bad cast is here:
https://github.com/AMReX-Codes/amrex/blob/0ca401daa57fd80e7ae7a12eb5fd28b9d10ce11b/Src/EB/AMReX_algoim.cpp#L25

(With the last push it might now be one call earlier.)

WeiqunZhang commented:
That's right.  I added a test before calling that function and that should also work.  The CI now fails for a different reason.

ax3l commented:
Is it possible the EB unit tests are pretty large in terms of runtime? I see the CI on ICC crunching on it for 20+ min

WeiqunZhang commented:
Yes, some EB tests are too large for CI.

ax3l commented:
Those are the biggest runtime eaters:
```
30/38 Test #30: CNS_Combustor .......................   Passed  145.31 sec
      Start 31: CNS_Pulse
32/38 Test #32: CNS_ShockRef ........................   Passed  4259.57 sec
      Start 33: CNS_Sod
33/38 Test #33: CNS_Sod .............................   Passed  112.03 sec
```

@WeiqunZhang can we just reduce their resolution and steps for the checked in inputs?
If they need to be large they might be better candidates for tutorials than the tests in the AMReX repo?

WeiqunZhang commented:
Yes, we can try to reduce the size or exclude some inputs files.

I also want to look at the EB coverage.  We already have some CI tests for EB in development.  I think we are converting too many non-EB tests to EB now.

ax3l commented:
For this PR to finish, shall I exclude all `CNS*` tests for now and we reduce in a follow-up?

ax3l committed:
CI Tests: Exclude CNS for now.

ax3l commented:
Excluded the `CNS_*` test in the last commit from CI.
Let's re-enable them in #2237 once we reduce the inputs.

ax3l commented:
@WeiqunZhang ready for merge :)

WeiqunZhang commented:
In AMReX, the EB codes are either put inside `#ifdef AMREX_USE_EB` or excluded in the build so that non-EB builds could work.  The issue in this PR is now the majority of the tests has enabled EB.  So, the CI tests may not be able to catch non-EB build errors.  Currently in the development branch, we are already testing EB builds in 4 GNU make CI tests with a combination of 2d, 3d, mpi, omp, hip, and cuda.  We want to add more coverage especially with cmake, but I think we should still leave most of the CI tests as non-EB builds.

ax3l committed:
Vary EB Tests

Keep a few tests without EB to ensure buildability.
Run tests whenever possible.

ax3l committed:
Mitigate new but problematic tests

ax3l commented on .github/workflows/linux.yml:
```diff
@@ -143,6 +153,8 @@ jobs:
             -DCMAKE_Fortran_COMPILER=$(which gfortran-10)
         make -j 2
 
+        ctest --output-on-failure -E Particles_GhostsAndVirtuals
```
Those should become addressed with #2247 (issue: #2243).
We can comment those in again once the linked PR is merged

ax3l commented:
@WeiqunZhang green now :)

WeiqunZhang committed:
have at least one non-EB test for Intel and Windows
